### ISSUE TITLE
AsyncOperation must be thread-safe

### ISSUE THREAD

**Issue by ole** [radianttap/Swift-Essentials#4]
I read [your article about `AsyncOperation`](http://aplus.rs/2018/asynchronous-operation/).

I believe your implementation isn't 100 % correct because the `state` property — and by extension the `isExecuting`, `isReady`, and `isFinished` properties — aren't thread-safe.

[The documentation for `Operation`](https://developer.apple.com/documentation/foundation/operation) states:

> When you subclass `NSOperation`, you must make sure that any overridden methods remain safe to call from multiple threads.

This isn't just a theoretical concern. I used to have a very similar implementation at first and when I ran it the thread sanitizer diagnosed tons of data races. You should protect all accesses to the `state` property behind a mutex, dispatch queue, or similar.

**Comment by ole** [comment]
I do it by giving the class a private serial `DispatchQueue`. All stored properties (I believe you only have `state`) must only be accessed through that queue. The easiest way to ensure this is to provide a computed property that everyone should use (written from memory, might not compile):

```swift
open class AsyncOperation : Operation {
    ...
    /// Mutex for all stored properties of this class
    private let queue = DispatchQueue(label: "AsyncOperation")
    /// Must only be accessed on queue
    private var _state = State.ready {
		willSet {
			willChangeValue(forKey: _state.key)
			willChangeValue(forKey: newValue.key)
		}
		didSet {
			didChangeValue(forKey: oldValue.key)
			didChangeValue(forKey: _state.key)
		}
	}

    /// Everyone should use this property as the interface
    public private(set) state: State {
        get { return queue.sync { _state } }
        set { queue.sync { _state = newValue } }
    }

    ...
}
```

**Comment by radianttap** [comment]
Tested with TSan in my async-heavy live project and sure enough, here it is.

<img width="1345" alt="tsan" src="https://user-images.githubusercontent.com/80697/43466862-c5606c7c-94e0-11e8-8a08-e00416838173.png">

**Comment by radianttap** [comment]
Tested the same large project with the updated version of the AsyncOperation and TSan is not reporting any issues. 🤞🏻

**Comment by ole** [comment]
Great. 👍

**Comment by aodhol** [comment]
I'm still seeing occasional TSan issues - do you need to synchronize access to the getter also?

**Comment by radianttap** [comment]
I can’t repeat that. Any ideas what might cause it?

**Comment by aodhol** [comment]
@radianttap I can reliably reproduce the TSan issue but only in a project i.e. no sign of anything going wrong when using a playground (perhaps obviously). Please copy the code here into a real project: https://github.com/aodhol/RaceCondition

Thanks!

**Comment by ole** [comment]
Here's the [Thread Sanitizer documentation on data races](https://developer.apple.com/documentation/code_diagnostics/thread_sanitizer/data_races):

> This check detects when multiple threads access the same memory without synchronization **and at least one access is a write.**

So yeah, the getter should absolutely be protected by the queue, too (which it isn't anymore in current master).

That said, I think my version is not 100% "correct" either. For context, see [ReactiveSwift/#269](https://github.com/ReactiveCocoa/ReactiveSwift/issues/269) (it's about subscripts, but I think getters/setters work the same way) and [this diff from @objcio](https://github.com/objcio/S01E42-thread-safety-reactive-programming-5/commit/2c8b4c60e2154776b575ce7641b6e23e4e8be12d).

If you use `AsyncOperation.state` as an l-value or in an `inout` context, Swift will call both the getter and setter, and this combination is not a single atomic operation in our case. This is probably not a problem in practice and I'm not sure the Thread Sanitizer would catch it (because every single read and write access is properly locked)

But you could work around the potential issue by removing the setter for `state` and providing an "atomic modify" method similar to this:

open class AsyncOperation : Operation {
    ...
    /// Mutex for all stored properties of this class
    private let queue = DispatchQueue(label: "AsyncOperation")
    /// Must only be accessed on queue
    private var _state = State.ready {
        willSet { ... }
        didSet { ... }
    }

    /// Everyone should use this property as the interface
    // New: now immutable
    public state: State {
        return queue.sync { _state }
    }

    func modifyState(_ transform: (inout State) -> ()) {
        queue.sync {
            transform(&_state)
        }
    }

    ...
}

All mutations of `_state` would then have to go through `modifyState`. As I said, it's probably not strictly necessary in this particular case because `state` is unlikely to be used in an l-value context, but it shouldn't hurt either.

**Comment by radianttap** [comment]
Thanks for the effort guys; I’m on vacation so will take a look as soon as I can.

**Comment by ole** [comment]
I came across another problem. In your design you send the KVO notifications when `_state` is mutated:

```swift
private var _state = State.ready {
    willSet {
        willChangeValue(forKey: _state.key)
        willChangeValue(forKey: newValue.key)
    }
    didSet {
        didChangeValue(forKey: oldValue.key)
        didChangeValue(forKey: _state.key)
    }
}
```

When you now protect the getter with the isolation queue, i.e.:

```swift
    public private(set) state: State {
        get { return queue.sync { _state } }
        set { queue.sync { _state = newValue } }
    }
```

You'll notice that this code deadlocks when you use these operations because sending `didChangeValue` causes KVO observers to read the state via calls to `isReady`, `isExecuting` and `isFinished`. And this causes a deadlock because we're indirectly calling `queue.sync` inside another `queue.sync` block.

I did not have this problem in my implementation because I sent the KVO notifications outside the queue. The problem with that approach is that you can't send as fine-grained notifications as you are doing here. I would essentially send `willChange`/`didChange` pairs for all potentially affected keys (`isReady`, `isExecuting`, `isFinished`) on every state mutation. For example, a KVO notification on `isFinished` would fire (with an unchanged value) when the state changed from `isReady` to `isExecuting`.

I like the more fine-grained approach you have, but implementing it correctly seems difficult. We need to do these things, in order:

1. Read the old state value before the mutation (on `queue`).
2. Send `willChange` (outside `queue`).
3. Perform the mutation (on `queue`).
4. Send `didChange` (outside `queue).

Implementing it in this way opens the door for synchronization issues when one thread is interrupted by another thread e.g. after step 2, but before step 3. I'm not yet sure if this would cause problems, but it's definitely a code smell. You could work around it by wrapping the entire procedure (steps 1-4) inside another serial queue. That seems inelegant too, but it may be the best solution?

**Comment by ole** [comment]
Here's my latest `AsyncOperation` version: https://gist.github.com/ole/5034ce19c62d248018581b1db0eabb2b. This uses an additional queue to solve the problem I outlined in the previous post. I'm pretty happy about it.

I don't think the additional queue switching incurs any significant overhead. GCD should be smart enough to run all queues on the same thread (I tested this informally with print statements, though that's not a guarantee, of course).
